plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,537 @@
+/*
+ * Copyright (C) 2023 Algoltek, Inc.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-algoltek-usb-common.h"
+#include "fu-algoltek-usb-device.h"
+#include "fu-algoltek-usb-firmware.h"
+#include "fu-algoltek-usb-struct.h"
+
+struct _FuAlgoltekUsbDevice {
+	FuUsbDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuAlgoltekUsbDevice, fu_algoltek_usb_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+algoltek_device_ctrl_transfer(FuAlgoltekUsbDevice *self,
+			      GUsbDeviceDirection direction,
+			      guint8 algoltek_cmd,
+			      guint16 value,
+			      guint16 index,
+			      GByteArray *transfer_pkt,
+			      guint8 len,
+			      GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   direction,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   algoltek_cmd,
+					   value,
+					   index,
+					   transfer_pkt->data,
+					   len,
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "Device communication failure: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static GByteArray *
+algoltek_device_rdr(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 5;
+	buf->data[1] = FU_ALGOLTEK_CMD_RDR;
+	fu_memwrite_uint16(buf->data + 2, address, G_BIG_ENDIAN);
+	buf->data[10] = ~fu_sum8(buf->data, buf->len) + 1;
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   FU_ALGOLTEK_CMD_RDR,
+					   address,
+					   0xFFFF,
+					   buf,
+					   buf->len,
+					   error))
+		return NULL;
+
+	/* success */
+	return g_steal_pointer(&buf);
+}
+
+static GByteArray *
+algoltek_device_rdv(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GByteArray) versionData = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 64, 0x0);
+	buf->data[0] = 3;
+	buf->data[1] = FU_ALGOLTEK_CMD_RDV;
+	buf->data[63] = ~fu_sum8(buf->data, buf->len) + 1;
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   FU_ALGOLTEK_CMD_RDV,
+					   0xFFFF,
+					   0xFFFF,
+					   buf,
+					   buf->len,
+					   error))
+		return NULL;
+
+	/* Remove len and cmd bytes */
+	for (guint32 i = 2; i < buf->len; i++) {
+		if (buf->data[i] < 128)
+			fu_byte_array_append_uint8(versionData, buf->data[i]);
+	}
+	/* success */
+	return g_steal_pointer(&versionData);
+}
+
+static guint16
+fu_algoltek_readout_value(GByteArray *back_data)
+{
+	guint16 back_data_addr;
+	guint16 back_data_value;
+
+	back_data_addr = fu_memread_uint16(back_data->data + 2, G_BIG_ENDIAN);
```

Comment:
I think fu_memread_uint16_safe() might be a good idea here. And then adding `GError **` to `fu_algoltek_readout_value()`

Reply:
Updated!